### DISCUSSION TITLE
RecordArray masks differently than normal Array

### DISCUSSION THREAD

**Discussion by lnoehte** [scikit-hep/awkward#1585] [Q&A] [answered]
Hi,
I just ran into something unexpected when masking a RecordArray. If I have to mask RecordArrays differently compared to normal Arrays, let me know. I didn't find anything about that, yet.
I am using awkward version 1.8.0.

To give an example which behavior I would expect, I will mask an Array first:

```python
import awkward as ak
array1 = ak.Array(
    [
        [53.1, 30.2],
        [],
        [91.7, 88.6, 20.3],
    ]
)
print("array1:")
print(array1)

mask1 = array1 > 40
print("mask1:")
print(mask1)

print()
print("slicing array1 with mask1:")
print(array1[mask1])
print("masking array1 with mask1:")
print(array1.mask[mask1])
print(ak.mask(array1, mask1))
```

The output is the following:

```python
array1:
[[53.1, 30.2], [], [91.7, 88.6, 20.3]]
mask1:
[[True, False], [], [True, True, False]]

slicing array1 with mask1:
[[53.1], [], [91.7, 88.6]]
masking array1 with mask1:
[[53.1, None], [], [91.7, 88.6, None]]
[[53.1, None], [], [91.7, 88.6, None]]
```

This is exactly what I was expecting to get.

Now if I do the same thing with a RecordArray:

```python
array2 = ak.Array(
    [
        {
            'Muon_pt': [53.1, 30.2],
        },
        {
            'Muon_pt': [],
        },
        {
            'Muon_pt': [91.7, 88.6, 20.3],
        }
    ]
)
print("array2:")
print(array2)
print(array2.tolist())

mask2 = array2.Muon_pt > 40
print("mask2:")
print(mask2)

print()
print("slicing array2 with mask2:")
print(array2[mask2])
print("masking array2 with mask2:")
print(array2.mask[mask2])
print(ak.mask(array2, mask2))

print()
print("to list:")
print()
print("slicing array2 with mask2:")
print(array2[mask2].tolist())
print("masking array2 with mask2:")
print(array2.mask[mask2].tolist())
print(ak.mask(array2, mask2).tolist())
```

The output is:
```python
array2:
[{Muon_pt: [53.1, 30.2]}, {Muon_pt: []}, {Muon_pt: [91.7, 88.6, 20.3]}]
[{'Muon_pt': [53.1, 30.2]}, {'Muon_pt': []}, {'Muon_pt': [91.7, 88.6, 20.3]}]
mask2:
[[True, False], [], [True, True, False]]

slicing array2 with mask2:
[{Muon_pt: [53.1]}, {Muon_pt: []}, {Muon_pt: [91.7, 88.6]}]
masking array2 with mask2:
[[{Muon_pt: [53.1, 30.2]}, None], ... 20.3]}, {Muon_pt: [91.7, 88.6, 20.3]}, None]]
[[{Muon_pt: [53.1, 30.2]}, None], ... 20.3]}, {Muon_pt: [91.7, 88.6, 20.3]}, None]]

to list:

slicing array2 with mask2:
[{'Muon_pt': [53.1]}, {'Muon_pt': []}, {'Muon_pt': [91.7, 88.6]}]
masking array2 with mask2:
[[{'Muon_pt': [53.1, 30.2]}, None], [], [{'Muon_pt': [91.7, 88.6, 20.3]}, {'Muon_pt': [91.7, 88.6, 20.3]}, None]]
[[{'Muon_pt': [53.1, 30.2]}, None], [], [{'Muon_pt': [91.7, 88.6, 20.3]}, {'Muon_pt': [91.7, 88.6, 20.3]}, None]]
```

But what I was expecting was:

```python
[{'Muon_pt': [53.1, None]}, {'Muon_pt': []}, {'Muon_pt': [91.7, 88.6, None]}]
```

What I see is, that the shape of array2 changed and is now the shape of mask2. I would expect ak.mask to preserve the shape of the initial array.

I don't know if this is the normal behavior of RecordArrays but I would be happy to find a way to get the expected output.

Thanks for all comments and suggestions.

**Comment by agoose77** [comment]
What's happening here is that the mask and the array are first broadcast together. In this case, this changes the type of the array:

```python3
>>> array2.type
3 * {"Muon_pt": var * float64}

>>> mask2.type
3 * var * bool

>>> ak.broadcast_arrays(array2,mask2)
[<Array [[{Muon_pt: [53.1, 30.2, ... 20.3]}]] type='3 * var * {"Muon_pt": var * f...'>,
 <Array [[{Muon_pt: [True, True, ... False]}]] type='3 * var * {"Muon_pt": var * ...'>]
```

It's not so obvious why `array2.mask[mask2]` doesn't produce the result that you expect until you consider what would happen if your record had more than a single field. In that case, if e.g. it had a list field called `x`, how should this be masked if `x` had a different shape to `Muon_pt`?

More specifically, you need to provide Awkward with your mask in a RecordArray, e.g. 
```python
mask2 = ak.zip({
   "Muon_pt": array2.Muon_pt > 40
}, depth_limit=1)
```
This will ensure that the broadcast result matches what you expect to happen.

Whether this is the right approach for your problem, though, I can't say! :)

**Reply by agoose77** [reply]
Pinging @jpivarski because I feel like there's more to add here, but *right now* my brain is not 100% working!

**Reply by jpivarski** [reply]
This looks good! Sorry that I was too busy to help out earlier.

**Comment by lnoehte** [comment, marked as answer]
Thanks a lot for the quick response.

When I saw the shape of `mask2` in `array2` I already suspected something happening when broadcasting, that I didn't expect.

I also tried it with more than one field, like

```python
array2 = ak.Array(
    [
        {
            'Muon_pt': [53.1, 30.2],
            'Muon_eta': [1.1, 0.2],
        },
        {
            'Muon_pt': [],
            'Muon_eta': [],
        },
        {
            'Muon_pt': [91.7, 88.6, 20.3],
            'Muon_eta': [0.6, -0.5],
        }
    ]
)
```

I tried to keep the first example simple, because it seems to yield the same result:

```python
array2:
[{Muon_pt: [53.1, 30.2], Muon_eta: [1.1, 0.2], ... 20.3], Muon_eta: [0.6, -0.5]}]
[{'Muon_pt': [53.1, 30.2], 'Muon_eta': [1.1, 0.2]}, {'Muon_pt': [], 'Muon_eta': []}, {'Muon_pt': [91.7, 88.6, 20.3], 'Muon_eta': [0.6, -0.5]}]
mask2:
[[True, False], [], [True, True, False]]

slicing array2 with mask2:
[{Muon_pt: [53.1], Muon_eta: [1.1]}, ... 91.7, 88.6], Muon_eta: [0.6, -0.5]}]
masking array2 with mask2:
[[{Muon_pt: [53.1, 30.2], Muon_eta: [1.1, 0.2], ... Muon_eta: [0.6, -0.5]}, None]]
[[{Muon_pt: [53.1, 30.2], Muon_eta: [1.1, 0.2], ... Muon_eta: [0.6, -0.5]}, None]]

to list:

slicing array2 with mask2:
[{'Muon_pt': [53.1], 'Muon_eta': [1.1]}, {'Muon_pt': [], 'Muon_eta': []}, {'Muon_pt': [91.7, 88.6], 'Muon_eta': [0.6, -0.5]}]
masking array2 with mask2:
[[{'Muon_pt': [53.1, 30.2], 'Muon_eta': [1.1, 0.2]}, None], [], [{'Muon_pt': [91.7, 88.6, 20.3], 'Muon_eta': [0.6, -0.5]}, {'Muon_pt': [91.7, 88.6, 20.3], 'Muon_eta': [0.6, -0.5]}, None]]
[[{'Muon_pt': [53.1, 30.2], 'Muon_eta': [1.1, 0.2]}, None], [], [{'Muon_pt': [91.7, 88.6, 20.3], 'Muon_eta': [0.6, -0.5]}, {'Muon_pt': [91.7, 88.6, 20.3], 'Muon_eta': [0.6, -0.5]}, None]]
```
But ultimately, I have a bunch of fields in that RecordArray.

Thanks a lot for the example using `ak.zip`.
If I see it correctly, this only works if I have only the "Muon_pt" field.
I quickly tried to make it work with the new `array2` from above by having two identical fields with the keys `Muon_pt` and `Muon_eta` in the mask but till now I was unsuccessful with that approach.

I am also wondering if I really need to have duplicates of this `array2.Muon_pt > 40` array for each field that exists in `array2` in the `mask2`, considering that I actually have a few more than just two.

**Reply by agoose77** [reply]
@lnoehte it depends upon what you're trying to do here. The main issue here is that your record structures are not deeply nested, i.e. your array has type
```python
3 * {"Muon_pt": var * float64, "Muon_eta": var * float64}
```
instead of 
```python
3 * var * {"Muon_pt": float64, "Muon_eta": float64}
```

The form of your array is an important choice that should make it easy to work with your data in a natural way. In the case of records, that means choosing whether to have `var * {'x': float6}` or `{'x': var * float64}`. In this instance, it feels like you want the former, not the latter. Otherwise, you need to slice each field, which is what we discussed above :)

**Reply by lnoehte** [reply]
Thanks a lot for the reply.
I believe it helped a lot for my understanding.

Maybe I should explain my situation.
I read a ROOT file with flat Ntuple like structure and I load branches with the `.arrays()` method. What I get after reading a subset of the branches is:
```python
1089694 * {"Muon_pt": var * float32, "Muon_eta": var * float32, "Muon_phi": var * float32, "Muon_tightId": var * bool, "Muon_pfRelIso04_all": var * float32, "Muon_charge": var * int32}
```
So I don't have any influence in the original data structure. But I just found that using `how="zip"` in the `.arrays()` method gives me:
```python
1089694 * {"Muon": var * {"pt": float32, "eta": float32, "phi": float32, "tightId": bool, "pfRelIso04_all": float32, "charge": int32}}
```
which is maybe a step in the right direction. 

I will keep testing the next days.

**Reply by agoose77** [reply]
Ah, OK. This is commonly seen with reading ROOT files - the level of jaggedness isn't correlated by default between branches, so the record is added at a higher depth. You can zip together branches manually, and set the appropriate `depth_limit` if you want. Or, you can pass `how="zip"` to `TTree.arrays()` from uproot, and uproot will do this for you.

**Reply by lnoehte** [reply]
Due to other duties I didn't find enough time to follow up on this the last weeks.
I wanted to continue the next days investigating how to mask my RecordArrays properly.

@agoose77 Could you maybe help me out and give an example of how the RecordArray structure/layout should look like, to get the behavior I see when slicing, where broadcasting doesn't seem to be an issue, so I know what to aim for when trying different ways to load my input file with uproot.

**Reply by agoose77** [reply]
In my [above comment](https://github.com/scikit-hep/awkward/discussions/1585#discussioncomment-3397240), the difference between a record of arrays and array of records is:

Record of arrays:
```python
3 * {"Muon_pt": var * float64, "Muon_eta": var * float64}
```

Array of records:
```python
3 * var * {"Muon_pt": float64, "Muon_eta": float64}
```

Uproot returns the "record of arrays", as by default there is no information about whether the branches are compatible with one another.

You can use pass `how="zip"` into the `arrays()` method that `uproot` provides to instruct it to give you the branches as a zipped array. You might need to limit `arrays()` to only the branches that you want to zip together. A total example:

```python
events = uproot.open(...)

momentum = events.arrays(["Muon_pt", "Muon_eta"], how="zip")
```

**Reply by lnoehte** [reply]
It took a while again but I created some tests.
I decided to prepare a test file to show my test results with the exact same structure as my input file.

Unfortunately both array types don't broadcast properly. It is still unclear to me why broadcasting works for slicing but not for masking. It doesn't look to me anymore as if it was an issue with the type, since slicing works in all cases we discussed above.

To show what's happening with the arrays for the different array types, I prepared a notebook, which you can find here:
[https://mybinder.org/v2/gh/lnoehte/test_uproot_awkward_with_nanoAOD.git/main](https://mybinder.org/v2/gh/lnoehte/test_uproot_awkward_with_nanoAOD.git/main)

I would be curious to see an example with `ak.mask` that works with RecordArray structures, that have more than one field.

Thanks a lot in advance again.

**Reply by agoose77** [reply]
@lnoehte thank you for the MyBinder example, it helps to clarify your expectations. 

Your "Automatic zip → Masking" example is another instance of https://github.com/scikit-hep/awkward/discussions/1585#discussioncomment-3386425 

Records in Awkward make it possible for arrays to have dimensions that are not defined by a single number, i.e. each field of an array could be an array of different dimension. Therefore, for the most part, records behave like "atoms" in Awkward; many operations stop when they hit a record.

Here's an excerpt of your data:
```python
import awkward as ak
import numpy as np

array = ak.from_iter(
    [
        {
            "Muon": [
                {
                    "pt": 53.02759552001953,
                    "eta": 1.1298828125,
                    "phi": 1.251220703125,
                    "tightId": True,
                    "pfRelIso04_all": 0.018847892060875893,
                    "charge": -1,
                },
                {
                    "pt": 30.640729904174805,
                    "eta": 0.162200927734375,
                    "phi": -2.2509765625,
                    "tightId": True,
                    "pfRelIso04_all": 0.016491081565618515,
                    "charge": 1,
                },
            ]
        },
        {"Muon": []},
        {
            "Muon": [
                {
                    "pt": 91.78971099853516,
                    "eta": 0.66748046875,
                    "phi": -1.208251953125,
                    "tightId": True,
                    "pfRelIso04_all": 0.0,
                    "charge": -1,
                },
                {
                    "pt": 88.6688232421875,
                    "eta": -0.584716796875,
                    "phi": 1.92626953125,
                    "tightId": True,
                    "pfRelIso04_all": 0.005749838892370462,
                    "charge": 1,
                },
            ]
        },
        {
            "Muon": [
                {
                    "pt": 36.98743438720703,
                    "eta": -0.3990478515625,
                    "phi": -1.376708984375,
                    "tightId": True,
                    "pfRelIso04_all": 0.0,
                    "charge": 1,
                },
                {
                    "pt": 32.94463348388672,
                    "eta": -2.0595703125,
                    "phi": 1.7509765625,
                    "tightId": True,
                    "pfRelIso04_all": 0.08915365487337112,
                    "charge": -1,
                },
            ]
        },
        {
            "Muon": [
                {
                    "pt": 16.899534225463867,
                    "eta": 2.24853515625,
                    "phi": -1.506591796875,
                    "tightId": False,
                    "pfRelIso04_all": 0.20170965790748596,
                    "charge": -1,
                }
            ]
        },
        {
            "Muon": [
                {
                    "pt": 42.16910171508789,
                    "eta": 2.115234375,
                    "phi": 0.918701171875,
                    "tightId": True,
                    "pfRelIso04_all": 0.046248190104961395,
                    "charge": 1,
                }
            ]
        },
        {"Muon": []},
        {
            "Muon": [
                {
                    "pt": 3.306669235229492,
                    "eta": 0.9600830078125,
                    "phi": -0.5576171875,
                    "tightId": False,
                    "pfRelIso04_all": 0.11646004021167755,
                    "charge": -1,
                }
            ]
        },
        {
            "Muon": [
                {
                    "pt": 3.9849512577056885,
                    "eta": 1.9267578125,
                    "phi": -2.970703125,
                    "tightId": False,
                    "pfRelIso04_all": 2.214714288711548,
                    "charge": 1,
                }
            ]
        },
        {"Muon": []},
    ]
)

```

If we look at the type, we can see that there's an outermost "Muon" record:
```python
>>> array.type
10 * {
    Muon: var * {
        pt: float64,
        eta: float64,
        phi: float64,
        tightId: bool,
        pfRelIso04_all: float64,
        charge: int64
    }
}
```

Therefore, it's not surprising that this array has dimension 1:
```python
>>> array.ndim
1
```

Looking at your mask, which is
```python
>>> array.Muon.eta.type # ignore what we do to this, just care about dimensionality
10 * var * float64
```

we see that it has two dimensions. Therefore, when the array and the mask are broadcast against one-another, we'll use left-broadcasting to make the dimensions agree, i.e.
```
10 * var * float64
10 * var * {Muon: var ...}
        ^^^ 
        new dimension added by broadcasting
```
It therefore makes sense that the result of the mask has this new dimension:
```python
>>> array.mask[array.Muon.pt > 0].type
10 * var * ?{
    Muon: var * {
        pt: float64,
        eta: float64,
        phi: float64,
        tightId: bool,
        pfRelIso04_all: float64,
        charge: int64
    }
}
```

You need to decide whether you want to mask the eta values _within_ the records, or whether you want to mask the records themselves. If you want to mask _within_ records (i.e. only touch array.Muon.eta, not array.Muon.phi) then you need to unzip, mask, and rezip the array. If you want to mask the record itself (which I believe you do), then the easiest thing to do is to pull the `Muon` field out of the array:
```python
masked_muon = array.Muon.mask[array.Muon.eta > 0]
```

**Reply by lnoehte** [reply]
That explanation makes sense to me.

Indeed, if I mask in eta, I also want to mask the same entries in phi and all the other fields, since they belong to the same particle.

Because of the "Muon" field around the array, I also showed the method when zipping manually.
This should already behave like pulling out the Muon field from your last example. Please correct me if I am mistaken. Unfortunately it still doesn't mask correctly but shows broadcast issues as well.

**Reply by agoose77** [reply]
Your "manual zip" example looks as I would expect it - you're correctly masking out the _records_ (`['pt', 'eta', 'phi', 'tightId', 'pfRelIso04_all', 'charge']`) according to the value of `eta`. The option type masks-out those records that do not satisfy your mask.

The difference between slicing and masking is subtle - well observed! I realise that you looked into this in your Binder example; I was looking at this quite late last night, so I focussed on the particular challenge of masking.

Ultimately slicing and masking follow different paths. The masking code broadcasts the mask against the array, which stops at the `Muon` record as discussed above. Meanwhile, the _slicing_ path doesn't use broadcasting; it traverses into the array being sliced, following a set of rules. 

At this point, I will ping @jpivarski to clarify our intentions here, because slicing and masking are highly similar operations, and I wonder whether this discrepancy is a good thing. I feel like we've covered this topic before, but I couldn't find anything in our issue tracker. Jim – do we _want_ slicing and masking to be different here? Is it intentional that the mask to stop at the first record, or is it just an artifact of stopping _broadcasting_ at records? i.e. do we need to change this, or is it deliberate?

**Reply by lnoehte** [reply]
No worries, I've seen that is was quite late yesterday. I am super grateful to get so much support here.

Concerning the "manual zip", it masks out the entries. I just realized that I might have read the structure incorrectly there and though it added a `None` at the wrong place when it might have done the right thing.

I'll take another look at the "manual zip" method and come back to you later today. I hope with success. If not, probably with more questions ;)

**Reply by lnoehte** [reply]
OK, I believe it looks fine to me as well.
I must have been confused from all the curly brackets. :D

One small question about the type. What does the `?` mean? Is it a sign that it's a mixed type because of the `None` added?

**Reply by agoose77** [reply]
`?xxx` is the same as `option[xxx]`. It's just easier to read than `option[...]`, so we use it where possible. The type grammar is heavily based upon [datashape](https://datashape.readthedocs.io/en/latest/overview.html#option)

**Reply by lnoehte** [reply]
Thanks a lot.
I cannot mark replies as answers.
Despite it being my "answer" (more a question), I'll mark this one here, since your first reply basically contains the answer to this discussion. I hope people will find it there.
